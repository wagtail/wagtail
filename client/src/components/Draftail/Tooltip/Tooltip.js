import PropTypes from 'prop-types';
import React from 'react';

const TOP = 'top';
const LEFT = 'left';
const TOP_LEFT = 'top-left';

const getTooltipStyles = (target, direction) => {
  const top = window.pageYOffset + target.top;
  const left = window.pageXOffset + target.left;
  switch (direction) {
    case TOP:
      return {
        top: top + target.height,
        left: left + target.width / 2,
      };
    case LEFT:
      return {
        top: top + target.height / 2,
        left: left + target.width,
      };
    case TOP_LEFT:
    default:
      return {
        top: top + target.height,
        left: left,
      };
  }
};

/**
 * A tooltip, with arbitrary content.
 */
const Tooltip = ({ target, children, direction }) => (
  <div
    style={getTooltipStyles(target, direction)}
    className={`Tooltip Tooltip--${direction}`}
    role="tooltip"
  >
    {children}
  </div>
);

Tooltip.propTypes = {
  target: PropTypes.shape({
    top: PropTypes.number.isRequired,
    left: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }).isRequired,
  direction: PropTypes.oneOf([TOP, LEFT, TOP_LEFT]).isRequired,
  children: PropTypes.node.isRequired,
};

export default Tooltip;
