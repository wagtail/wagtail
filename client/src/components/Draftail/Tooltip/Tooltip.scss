$tooltip-spacing: 0.625rem;
$tooltip-arrow-height: 15px;
$tooltip-arrow-spacing: 25px;
$tooltip-radius: $draftail-tooltip-radius;

$tooltip-chrome: $draftail-tooltip-chrome;
$tooltip-chrome-text: $draftail-tooltip-chrome-text;

$tooltip-z-index: $draftail-tooltip-z-index;

$tooltip-color-no: theme('colors.critical.100');

@mixin arrow--top {
  margin-top: $tooltip-arrow-spacing;
  transform: translateX(-50%);

  &::before {
    bottom: 100%;
    inset-inline-start: 50%;
    transform: translateX(-50%);
    border-bottom-color: $tooltip-chrome;
  }
}

@mixin arrow--left {
  margin-inline-start: $tooltip-arrow-spacing;
  transform: translateY(-50%);

  &::before {
    top: 50%;
    inset-inline-end: 100%;
    transform: translateY(-50%);
    border-inline-end-color: $tooltip-chrome;
  }
}

@mixin arrow--top-left {
  margin-top: $tooltip-arrow-spacing;

  &::before {
    bottom: 100%;
    inset-inline-start: $tooltip-arrow-spacing;
    border-bottom-color: $tooltip-chrome;
  }
}

.Tooltip {
  position: absolute;
  padding: $tooltip-spacing;
  background-color: $tooltip-chrome;
  color: $tooltip-chrome-text;
  z-index: $tooltip-z-index;
  border-radius: $tooltip-radius;
  box-shadow: 0 0 5px theme('colors.black-35');

  &::before {
    content: '';
    position: absolute;
    border: $tooltip-arrow-height solid transparent;
  }

  &--top {
    @include arrow--top;
  }

  &--left {
    @include arrow--left;
  }

  &--top-left {
    @include arrow--top-left;
  }

  &__link {
    @include font-smoothing;
    font-size: 0.875rem;
    margin-inline-end: $controls-spacing * 4;
    display: inline-block;

    &,
    &:hover {
      color: $color-white;
    }
  }

  &__button {
    &.no.button-secondary {
      color: $tooltip-color-no;
      border-color: currentColor;

      &:hover {
        color: $color-button-no;
      }
    }
  }
}
