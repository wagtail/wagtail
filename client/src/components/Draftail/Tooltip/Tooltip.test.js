import React from 'react';
import { shallow } from 'enzyme';

import Tooltip from './Tooltip';

const target = {
  top: 1,
  left: 1,
  width: 12,
  height: 1200,
};

describe('Tooltip', () => {
  it('#direction top', () => {
    expect(
      shallow(
        <Tooltip target={target} direction="top">
          Test
        </Tooltip>,
      ),
    ).toMatchSnapshot();
  });

  it('#direction left', () => {
    expect(
      shallow(
        <Tooltip target={target} direction="left">
          Test
        </Tooltip>,
      ),
    ).toMatchSnapshot();
  });

  it('#direction top-left', () => {
    expect(
      shallow(
        <Tooltip target={target} direction="top-left">
          Test
        </Tooltip>,
      ),
    ).toMatchSnapshot();
  });
});
