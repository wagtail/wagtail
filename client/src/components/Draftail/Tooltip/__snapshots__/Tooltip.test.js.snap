// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip #direction left 1`] = `
<div
  className="Tooltip Tooltip--left"
  role="tooltip"
  style={
    Object {
      "inset-inline-start": 13,
      "left": 13,
      "top": 601,
    }
  }
>
  Test
</div>
`;

exports[`Tooltip #direction top 1`] = `
<div
  className="Tooltip Tooltip--top"
  role="tooltip"
  style={
    Object {
      "inset-inline-start": 7,
      "left": 7,
      "top": 1201,
    }
  }
>
  Test
</div>
`;

exports[`Tooltip #direction top-left 1`] = `
<div
  className="Tooltip Tooltip--top-left"
  role="tooltip"
  style={
    Object {
      "inset-inline-start": 1,
      "left": 1,
      "top": 1201,
    }
  }
>
  Test
</div>
`;
